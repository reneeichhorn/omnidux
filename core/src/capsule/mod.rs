#[derive(Clone, Copy, Debug)]
pub enum CapsuleContent<T> where T: Copy + Sync + Send {
  Pending,
  Empty,
  Some(T),
}

impl<T: Copy + Sync + Send> CapsuleContent<T> {
  pub fn unwrap(&self) -> T {
    match self {
      CapsuleContent::Some(out) => *out,
      _ => panic!("Cannot unwrap CapsuleContent unless a value is returned."),
    }
  }
}

pub trait Capsule<K, V> where V: Copy + Sync + Send {
  fn request_content(&self, key: &K) -> CapsuleContent<V>;
  fn set_content(&self, key: K, value: CapsuleContent<V>);
}

#[macro_export]
macro_rules! capsule_get {
  ($context:ident, $repo:ident, $capsule:ident, $key:expr) => {
    {
      use std::any::Any;
      use omnidux_core::capsule::Capsule;

      let repo_uuid = unsafe { $repo::uuid };
      let repo = $context.get_repo(repo_uuid).as_any().downcast_ref::<$repo::Repository>().unwrap();
      repo.capsules.$capsule.request_content($key)
    }
  };
}

#[macro_export]
macro_rules! capsule_set {
  ($context:ident, $repo:ident, $capsule:ident, $key:expr, $value:expr) => {
    {
      use std::any::Any;
      use omnidux_core::capsule::Capsule;

      let repo_uuid = unsafe { $repo::uuid };
      let repo = $context.get_repo(repo_uuid).as_any().downcast_ref::<$repo::Repository>().unwrap();
      repo.capsules.$capsule.set_content($key, $value);
    }
  };
}

/// Implements the default capsule.
/// The default implementation is not safe when used within multiple threads.
#[macro_export]
macro_rules! impl_default_capsule {
  ($name: ident, $keyType: ident, $valueType: ident) => {
    pub struct $name {
      map: std::sync::RwLock<
        std::collections::HashMap<
          $keyType, CapsuleContent<$valueType>
        >
      >,
    }

    impl $name {
      fn new() -> $name {
        $name {
          map: std::sync::RwLock::new(std::collections::HashMap::new()),
        }
      }
    }

    impl omnidux_core::capsule::Capsule<$keyType, $valueType> for $name {
      fn request_content(&self, key: &$keyType) -> CapsuleContent<$valueType> {
        let map = self.map.read().unwrap();
        *map.get(key).or_else(|| Some(&CapsuleContent::Empty)).unwrap()
      }

      fn set_content(&self, key: $keyType, value: CapsuleContent<$valueType>) {
        let mut map = self.map.write().unwrap();
        map.insert(key, value);
      }
    }
  };
}
