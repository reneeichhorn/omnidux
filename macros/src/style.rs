pub use cssparser::{
  RuleListParser,
  ParserInput,
  Parser,
  Token,
  QualifiedRuleParser,
  BasicParseError,
  ParseError,
  AtRuleParser,
  CowRcStr,
  SourceLocation,
  DeclarationListParser,
  DeclarationParser,
};
use proc_macro2::{TokenStream, TokenTree, Literal};
use quote::{quote, format_ident, ToTokens};

pub struct StyleParser {}
pub struct CustomParser {}

#[derive(Debug)]
pub struct StyleDeclaration<'i> {
  pub name: String,
  pub token: Token<'i>,
}

#[derive(Debug)]
pub struct Rule<'i> {
  pub key: String,
  pub decls: Vec<StyleDeclaration<'i>>,
}

/// Some type information for our parser.
impl<'i> AtRuleParser<'i> for CustomParser {
  type PreludeBlock = ();
  type PreludeNoBlock = ();
  type AtRule = Rule<'i>;
  type Error = BasicParseError<'i>;
}

impl<'i> QualifiedRuleParser<'i> for CustomParser {
  type Prelude = String;
  type QualifiedRule = Rule<'i>;
  type Error = BasicParseError<'i>;

  /// Parses out the selector.
  fn parse_prelude<'t>(
      &mut self,
      input: &mut Parser<'i, 't>,
  ) -> Result<Self::Prelude, ParseError<'i, Self::Error>> {
      let location = input.current_source_location();

      let selector = match input.next()? {
        Token::Ident(ref element_name) => element_name.to_string(),
        t => { return Err(location.new_unexpected_token_error(t.clone())); }
      };

      Ok(selector)
  }

  /// Parses the block (`{...}`) into a Rule struct.
  fn parse_block<'t>(
    &mut self,
    key: Self::Prelude,
    _location: SourceLocation,
    input: &mut Parser<'i, 't>,
  ) -> Result<Self::QualifiedRule, ParseError<'i, Self::Error>> {
    let styles = DeclarationListParser::new(input, StyleParser {}).collect::<Vec<_>>();

    Ok(Rule {
      key: key,
      decls: styles.into_iter().filter_map(|decl| {
        if !decl.is_ok() {
          eprintln!("{:?}", decl);
        }

        decl.ok()
      }).collect()
    })
  }
}

impl<'i> AtRuleParser<'i> for StyleParser {
  type PreludeBlock = ();
  type PreludeNoBlock = ();
  type AtRule = StyleDeclaration<'i>;
  type Error = BasicParseError<'i>;
}

impl<'i> DeclarationParser<'i> for StyleParser {
    type Declaration = StyleDeclaration<'i>;
    type Error = BasicParseError<'i>;

    /// Parses a value (e.g, `background-color: #307ace;`) into a `Styles` value.
    fn parse_value<'t>(
      &mut self,
      name: CowRcStr<'i>,
      input: &mut Parser<'i, 't>,
    ) -> Result<Self::Declaration, ParseError<'i, Self::Error>> {
      Ok (StyleDeclaration { name: name.to_string(), token: input.next().unwrap().clone() })
    }
}

pub fn map_style_declaration<'i> (decl: &StyleDeclaration<'i>) -> TokenStream {
  fn map_dimension<'a>(token: Token<'a>) -> TokenStream {
    match token {
      Token::Number { value, .. } => TokenStream::from(TokenTree::Literal(Literal::f32_suffixed(value))),
      _ => panic!("Invalid style value")
    }
  }
  fn map_number<'a>(token: Token<'a>) -> TokenStream {
    match token {
      Token::Number { value, .. } => TokenStream::from(TokenTree::Literal(Literal::f32_suffixed(value))),
      _ => panic!("Invalid style value")
    }
  }

  let (name_str, value_appendix) = match (decl.name.as_ref()) {
    "flex-basis" => ("flex_basis", map_dimension(decl.token.clone())),
    "flex-grow" => ("flex_grow", map_number(decl.token.clone())),
    "flex-shrink" => ("flex_shrink", map_number(decl.token.clone())),
    _ => panic!("Unknown styling property")
  };

  let name = format_ident!("{}", name_str);

  quote! { #name: #value_appendix, }
}