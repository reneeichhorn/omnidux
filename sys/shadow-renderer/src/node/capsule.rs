use std::collections::HashMap;
use std::sync::RwLock;

use omnidux_core::capsule::{Capsule, CapsuleContent};
use stretch::Stretch;
use stretch::node::{Node as SNode};
use stretch::style::{Style as SStyles};
use stretch::geometry::{Size as SSize};

use super::node::{Node, NodeId, Layout};
use super::shadow::ShadowDiffer;
use crate::style::Styles;

pub struct NodeTree {
  stretch: RwLock<Stretch>,
  shadow: RwLock<HashMap<NodeId, Node>>,
  root: NodeId,
}

#[derive(Debug)]
pub enum ResolvedNode {
  Node { styles: Styles, layout: Layout, children: Vec<ResolvedNode> },
  Leaf,
}

impl NodeTree {
  /// Creates a new instance of the NodeTree and creates the first root node.
  pub fn new() -> NodeTree {
    let mut stretch = Stretch::new();

    let root = NodeId::from(stretch.new_node(Default::default(), Vec::new()).unwrap());
    let mut shadow = HashMap::new();
    shadow.insert(root, Node::new(Default::default()));

    NodeTree {
      stretch: RwLock::new(stretch),
      shadow: RwLock::new(shadow),
      root: root,
    }
  }

  /// Recursively travels the root node and resolves the whole tree (Debug purpose only)
  pub fn resolve(&self) -> ResolvedNode {
    self.resolve_node(self.root)
  }

  /// Recursively travels a node (Debug purpose only)
  pub fn resolve_node(&self, node: NodeId) -> ResolvedNode {
    let stretch = self.stretch.read().unwrap();
    let shadow = self.shadow.read().unwrap();
    let shadow_node = shadow.get(&node).unwrap();

    ResolvedNode::Node {
      styles: shadow_node.styles,
      layout: shadow_node.layout,
      children: stretch.children(SNode::from(node))
        .unwrap()
        .iter()
        .map(|child| self.resolve_node(NodeId::from(*child)))
        .collect(),
    }
  }
}

#[derive(Copy, Clone)]
pub enum NodeTreeKey {
  /// Targets all pending changes.
  PollPendingChanges,
  /// Targets the automatically create root node.
  RootNode,
  /// Targets a new node that belongs to a parent and contains a certain styling.
  NewNode(NodeId, Styles),
  /// Targets any existing node somewhere in the tree.
  ExNode(NodeId),
}

#[derive(Copy, Clone)]
pub enum NodeTreeAction {
  /// The response for a newly create node.
  Id(NodeId),
  /// The action that forces a node to be removed from the tree.
  Remove,
  /// The action that updates the styling of a tree.
  SetStyles(Styles),
}

impl NodeTreeAction {
  pub fn id (&self) -> NodeId {
    match self {
      NodeTreeAction::Id(id) => *id,
      _ => panic!("Node tree response was not a node id.")
    }
  }
}

impl Capsule<NodeTreeKey, NodeTreeAction> for NodeTree {
  /// Request is used to request the creation of a new empty node
  /// that can be later modified after rendering the content.
  fn request_content(&self, key: &NodeTreeKey) -> CapsuleContent<NodeTreeAction> {
    use NodeTreeKey::{*};

    match key {
      RootNode => {
        CapsuleContent::Some(NodeTreeAction::Id(self.root))
      },
      NewNode(parent, styles) => {
        // Create an empty node.
        let mut stretch = self.stretch.write().unwrap();
        let node = stretch.new_node(SStyles::from(*styles), Vec::new()).unwrap();

        // Append new node to parent.
        stretch.add_child(SNode::from(parent.clone()), node).unwrap();

        // Trigger recalculation
        stretch.compute_layout(SNode::from(self.root), SSize::undefined()).unwrap();

        {
          // Append shadow node.
          let mut shadow_tree = self.shadow.write().unwrap();
          shadow_tree.insert(NodeId::from(node), Node::new(*styles));

          // Synch stretch tree with shadow tree and find layout diffs.
          ShadowDiffer::new(&mut shadow_tree, &stretch)
            .with_layout(self.root);
        }

        // Return node id.
        CapsuleContent::Some(NodeTreeAction::Id(NodeId::from(node)))
      },
      _ => panic!("Requesting from capsule is only allowed for new content."),
    }
  }

  /// Can be used to update an existing node.
  fn set_content(&self, key: NodeTreeKey, value: CapsuleContent<NodeTreeAction>) {
    use NodeTreeKey::{*};
    use NodeTreeAction::{*};

    let node = match key {
      ExNode(node) => node,
      _ => panic!("Invalid NodeTree target."),
    };

    match value {
      CapsuleContent::Some(action) => {
        match action {
          SetStyles(styles) => {
            // Apply style to content.
            {
              let mut stretch = self.stretch.write().unwrap();
              stretch.set_style(SNode::from(node), SStyles::from(styles));

              // Trigger recalculation
              stretch.compute_layout(SNode::from(self.root), SSize::undefined()).unwrap();
            }

            // Apply new style to shadow tree.
            let mut shadow_tree = self.shadow.write().unwrap();
            shadow_tree.get_mut(&node).unwrap().styles = styles;

            // Synch stretch tree with shadow tree and find layout diffs.
            let stretch = self.stretch.read().unwrap();
            ShadowDiffer::new(&mut shadow_tree, &stretch)
              .with_layout(self.root);
          },
          _ => panic!("Invalid NodeTree action."),
        }
      },
      _ => panic!("NodeTree only supports CapsuleContent::some(..)"),
    }
  }
}