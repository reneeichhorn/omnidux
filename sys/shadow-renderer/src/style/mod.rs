use stretch::style::{Style as SStyles, Dimension as SDimension};
use stretch::geometry::{Size as SSize};

#[derive(Copy, Clone, Debug)]
pub enum Dimension {
  Undefined,
  Auto,
  Points(f32),
  Percent(f32),
}

#[derive(Copy, Clone, Debug)]
pub struct Rect<T> {
  pub left: T,
  pub right: T,
  pub top: T,
  pub bottom: T,
}

#[derive(Copy, Clone, Debug)]
pub struct Styles {
  pub flex_basis: Dimension,
  pub flex_grow: f32,
  pub flex_shrink: f32,
  pub width: Dimension,
  pub height: Dimension,
}

impl Default for Styles {
  fn default() -> Styles {
    Styles {
      flex_basis: Dimension::Auto,
      flex_grow: 0.0f32,
      flex_shrink: 1.0f32,
      width: Dimension::Auto,
      height: Dimension::Auto,
    }
  }
}

impl From<Dimension> for SDimension {
  fn from(d: Dimension) -> Self {
    match d {
      Dimension::Undefined => SDimension::Undefined,
      Dimension::Auto => SDimension::Auto,
      Dimension::Points(p) => SDimension::Points(p),
      Dimension::Percent(p) => SDimension::Percent(p),
    }
  }
}

impl From<Styles> for SStyles {
  fn from(styles: Styles) -> Self {
    SStyles {
      flex_basis: SDimension::from(styles.flex_basis),
      flex_grow: styles.flex_grow,
      flex_shrink: styles.flex_shrink,
      size: SSize {
        width: SDimension::from(styles.width),
        height: SDimension::from(styles.height),
      },
      ..Default::default()
    }
  }
}