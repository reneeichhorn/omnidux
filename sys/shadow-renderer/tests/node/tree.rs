extern crate insta;
extern crate omnidux_core;
extern crate omnidux_sys_shadow_renderer;

use insta::assert_debug_snapshot;
use omnidux_core::capsule::{Capsule, CapsuleContent};
use omnidux_sys_shadow_renderer::style::{Styles, Dimension};
use omnidux_sys_shadow_renderer::node::capsule::NodeTree;
use omnidux_sys_shadow_renderer::node::capsule::NodeTreeKey::*;
use omnidux_sys_shadow_renderer::node::capsule::NodeTreeAction::*;

#[test]
fn test_tree_mounting() {
  let nodeTree = NodeTree::new();

  let root = nodeTree.request_content(&RootNode).unwrap().id();
  nodeTree.set_content(ExNode(root), CapsuleContent::Some(SetStyles(Styles {
    width: Dimension::Points(600.0),
    height: Dimension::Points(400.0),
    ..Default::default()
  })));

  let flex1 = Styles {
    flex_basis: Dimension::Points(0.0),
    flex_grow: 1.0,
    flex_shrink: 1.0,
    ..Default::default()
  };

  let sub_1 = nodeTree.request_content(&NewNode(root, flex1));
  let sub_2 = nodeTree.request_content(&NewNode(root, flex1));

  let resolved = nodeTree.resolve();

  assert_debug_snapshot!(resolved);
}