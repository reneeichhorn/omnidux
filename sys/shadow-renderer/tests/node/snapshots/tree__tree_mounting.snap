---
source: tests/node/tree.rs
expression: resolved
---
Node {
    styles: Styles {
        flex_basis: Auto,
        flex_grow: 0.0,
        flex_shrink: 1.0,
        width: Points(
            600.0,
        ),
        height: Points(
            400.0,
        ),
    },
    layout: Layout {
        width: 600.0,
        height: 400.0,
        x: 0.0,
        y: 0.0,
    },
    children: [
        Node {
            styles: Styles {
                flex_basis: Points(
                    0.0,
                ),
                flex_grow: 1.0,
                flex_shrink: 1.0,
                width: Auto,
                height: Auto,
            },
            layout: Layout {
                width: 300.0,
                height: 400.0,
                x: 0.0,
                y: 0.0,
            },
            children: [],
        },
        Node {
            styles: Styles {
                flex_basis: Points(
                    0.0,
                ),
                flex_grow: 1.0,
                flex_shrink: 1.0,
                width: Auto,
                height: Auto,
            },
            layout: Layout {
                width: 300.0,
                height: 400.0,
                x: 300.0,
                y: 0.0,
            },
            children: [],
        },
    ],
}
